### PR TITLE
Less aggressive persistent reduction when it could induce large masking with dynamic shapes

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #163365

As per comment in source code:
```
            # If we are are coalescing on xblock (not ReductionHint.INNER) and this is not a tiny kernel
            # (not ReductionHint.OUTER_TINY), do not use persistent reduction if it induces tile
            # quantization. Peristent reduction forces rblock == rnumel, if the bounds between lower
            # and upper are large, for the lower values we will be masking off large % of read/writes,
            # when we could expand the coalescing xblock instead.
```

For the test case in question, this pr improves perf from 0.8573521325143717 -> 0.043151492193814305 because we were egregiously masking out rblock values (58/64 values). 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

Differential Revision: [D82853279](https://our.internmc.facebook.com/intern/diff/D82853279)

### CODE REVIEW

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D82853279).

File: test/inductor/test_torchinductor_dynamic_shapes.py [hunk 1]
```diff
@@ -1116,6 +1116,36 @@ def fn(x):
...
+        x = torch.randn(reduction_dim, batch, features, device=GPU_TYPE).permute(1, 2, 0)
+        y = torch.randn(reduction_dim, batch, features, device=GPU_TYPE).permute(1, 2, 0)
+
+        torch._dynamo.mark_dynamic(x, 2, min=6, max=64)
+        torch._dynamo.mark_dynamic(y, 2, min=6, max=64)
+
+        compiled_fn = torch.compile(reduce_bounded)
+        result, source_codes = run_and_get_code(compiled_fn, x, y)
+
+        FileCheck().check_not("@triton_heuristics.reduction").run(source_codes[0])
```
Comment (shunting314) [inline test/inductor/test_torchinductor_dynamic_shapes.py]: Did you actually want to check the opposite? Since persistent reduction is disabled in this case.

File: test/inductor/test_torchinductor_dynamic_shapes.py [hunk 2]
```diff
@@ -1116,6 +1116,36 @@ def fn(x):
...
+            result = torch.sum(x * y, dim=2)
+            return result
+
+        # Create tensors where reduction dimension is 6 (but could be up to 128)
+        batch = 256
+        features = 5536
+        reduction_dim = 6  # Actual size is small
+
+        x = torch.randn(reduction_dim, batch, features, device=GPU_TYPE).permute(1, 2, 0)
+        y = torch.randn(reduction_dim, batch, features, device=GPU_TYPE).permute(1, 2, 0)
```
Comment (shunting314) [inline test/inductor/test_torchinductor_dynamic_shapes.py]: not a big deal, but I think batch/features dimensions can be merged. So a 2D tensor here is enough.

**Review by shunting314 — APPROVED** [review]

**Review by PaulZhang12 — APPROVED** [review]
Nice!

**Review by jansel — APPROVED** [review]

**Review by v0i0 — APPROVED** [review]
